sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelper.cs

```diff
@@ -27,118 +27,77 @@
 using CS = Microsoft.CodeAnalysis.CSharp;
 using VB = Microsoft.CodeAnalysis.VisualBasic;
 
+using static Microsoft.CodeAnalysis.CSharp.LanguageVersion;
+using static Microsoft.CodeAnalysis.VisualBasic.LanguageVersion;
+using SonarAnalyzer.UnitTest.Helpers;
+
 namespace SonarAnalyzer.UnitTest.TestFramework
 {
+
+    /// <summary>
+    /// Returned values depend on the build environment.
+    /// Local run: Only the latest language version is used to simplify debugging.
+    /// CI run: All configured versions are returned.
+    /// </summary>
     internal static class ParseOptionsHelper
     {
-        private static readonly IEnumerable<ParseOptions> defaultParseOptions;
+        public static ImmutableArray<ParseOptions> BeforeCSharp7 { get; }
+        public static ImmutableArray<ParseOptions> BeforeCSharp8 { get; }
+
+        public static ImmutableArray<ParseOptions> FromCSharp6 { get; }
+        public static ImmutableArray<ParseOptions> FromCSharp7 { get; }
+        public static ImmutableArray<ParseOptions> FromCSharp8 { get; }
+
+        public static ImmutableArray<ParseOptions> FromVisualBasic12 { get; }
+        public static ImmutableArray<ParseOptions> FromVisualBasic14 { get; }
+        public static ImmutableArray<ParseOptions> FromVisualBasic15 { get; }
+
+        private static readonly ImmutableArray<ParseOptions> defaultParseOptions;
 
 #pragma warning disable S3963 // The static fields are dependent between them so the values cannot be set inline
+
         static ParseOptionsHelper()
         {
-            BeforeCSharp7 = CSharp5.Concat(CSharp6);
-            BeforeCSharp8 = BeforeCSharp7.Concat(CSharp7);
+            var cs7 = CreateOptions(CSharp7, CSharp7_1, CSharp7_2, CSharp7_3);
+            var vb15 = CreateOptions(VisualBasic15, VisualBasic15_3, VisualBasic15_5);
+
+            BeforeCSharp7 = CreateOptions(CSharp5).Concat(CreateOptions(CSharp6)).FilterByEnvironment();
+            BeforeCSharp8 = BeforeCSharp7.Concat(cs7).FilterByEnvironment();
 
-            FromCSharp8 = CSharp8;
-            FromCSharp7 = CSharp7.Concat(FromCSharp8);
-            FromCSharp6 = CSharp6.Concat(FromCSharp7);
+            FromCSharp8 = CreateOptions(CSharp8).FilterByEnvironment();
```

Comment:
As far as I can tell, when the code is run on local machines the full list of options will be create and then we will filter it and only take the last. It would be great to create it right from the beginning.

One option would be to have CreateOptions do also the filtering and take only the last item if needed. Something similar with:
`BeforeCSharp8 = CreateOptions(CSharp5, CSharp6, CSharp7, CSharp7_1, CSharp7_2, CSharp7_3);`
In this way the versions used are clear (no need to check other methods) and we will avoid multiple iterations.

Reply:
The create&filter is done only locally and only in tests. I'd keep it this way for simplicity, it's not expensive in this context. 

I considered the full-listing before doing this, but it would lead to situation like this in the future. And that would be hard to maintain and easy to make a typo.
```
BeforeCSharp8 = CreateOptions(CSharp5, CSharp6, CSharp7, CSharp7_1, CSharp7_2, CSharp7_3);
BeforeCSharp9 = CreateOptions(CSharp5, CSharp6, CSharp7, CSharp7_1, CSharp7_2, CSharp7_3, CSharp8);
BeforeCSharp10 = CreateOptions(CSharp5, CSharp6, CSharp7, CSharp7_1, CSharp7_2, CSharp7_3, CSharp8, CSharp9);
BeforeCSharp11 = CreateOptions(CSharp5, CSharp6, CSharp7, CSharp7_1, CSharp7_2, CSharp7_3, CSharp8, CSharp9, CSharp10);
```
and even worse (non-tabular)
```
FromCSharp11 = CreateOptions(CSharp11)
FromCSharp10 = CreateOptions(CSharp10, CSharp11)
FromCSharp9 = CreateOptions(CSharp9, CSharp10, CSharp11)
FromCSharp8 = CreateOptions(CSharp8, CSharp9, CSharp10, CSharp11)
FromCSharp7 = CreateOptions(CSharp7, CSharp7_1, CSharp7_2, CSharp7_3, CSharp8, CSharp9, CSharp10, CSharp11)
FromCSharp6 = CreateOptions(CSharp6, CSharp7, CSharp7_1, CSharp7_2, CSharp7_3, CSharp8, CSharp9, CSharp10, CSharp11);
```

Reply:
Ok. It's a matter of preference in the end.